### DISCUSSION TITLE
Rotated Bit Packed Glyph Array

### DISCUSSION THREAD

**Discussion by Xylopyrographer** [Xylopyrographer/STAC#50] [Q&A] [answered]
@mqshaw 
How're you and the family been keeping? All well I pray.

Been working on a STAC variant. The ATOM based device is a bit small to be seen from the stage to the back wall so I'm building a version with a larger display & an external WiFi antenna; code name STAC-LR. Should also be able to incorporate OTA updating and mDNS with this (which I'll back port to the STAC).

Anyway,  larger COTS displays use an 8 x 8 RGB LED matrix instead of the ATOM 5 x 5. I've redone the display routines to accommodate, but hit a bit of a bump regarding the glyph matrix. A bit more detail in the attached sketch, but basically the 8 x 8 glyphs are in bit-packed format. The bitmap for each row is packed into a single byte (uint8_t), for a total of 8 bytes to represent the glyph.

Wondering if you would have a routine or be able to craft one that would produce the run-time glyph array in bit-packed format rotated according to the STAC orientation?

Merry Christmas!
[Display8.zip](https://github.com/Xylopyrographer/STAC/files/10232630/Display8.zip)

**Comment by mqshaw** [comment]
Hi Rob,

Things are going well, getting ready for Christmas break. Looking forward to it.

Yes, I would be happy to write a routine for you. It sounds pretty simple, each byte has the 8 bits for that row packed, so you just need a routine to unpack the bits, rotate them and then repack them for each glyph.

Mark


From: Xylopyrographer ***@***.***>
Sent: Wednesday, December 14, 2022 5:21 PM
To: Xylopyrographer/STAC ***@***.***>
Cc: Mark Shaw ***@***.***>; Mention ***@***.***>
Subject: [Xylopyrographer/STAC] Rotated Bit Packed Glyph Array (Discussion #50)


@mqshaw<https://github.com/mqshaw>
How're you and the family been keeping? All well I pray.

Been working on a STAC variant. The ATOM based device is a bit small to be seen from the stage to the back wall so I'm building a version with a larger display & an external WiFi antenna; code name STAC-LR. Should also be able to incorporate OTA updating and mDNS with this (which I'll back port to the STAC).

Anyway, larger COTS displays use an 8 x 8 RGB LED matrix instead of the ATOM 5 x 5. I've redone the display routines to accommodate, but hit a bit if a wall regarding the glyph matrix. A bit more detail in the attached sketch, but basically the 8 x 8 glyphs are in bit-packed format. The bitmap for each row is packed into a single byte (uint8_t), for a total of 8 bytes to represent the glyph.

Wondering if you would have a routine or be able to craft one that would produce the run-time glyph array rotated according to the STAC orientation?

Merry Christmas!
Display8.zip<https://github.com/Xylopyrographer/STAC/files/10232630/Display8.zip>

**Comment by beamholder** [comment]
Mark: Exactly what’s needed indeed. 👍

No rush. Whenever you have the time. 

Cheers,
    Rob
===================================


On Dec 16, 2022, at 12:04 AM, Mark Shaw ***@***.***> wrote:

﻿
Hi Rob, 

Things are going well, getting ready for Christmas break. Looking forward to it. 

Yes, I would be happy to write a routine for you. It sounds pretty simple, each byte has the 8 bits for that row packed, so you just need a routine to unpack the bits, rotate them and then repack them for each glyph. 

Mark 


From: Xylopyrographer ***@***.***> 
Sent: Wednesday, December 14, 2022 5:21 PM 
To: Xylopyrographer/STAC ***@***.***> 
Cc: Mark Shaw ***@***.***>; Mention ***@***.***> 
Subject: [Xylopyrographer/STAC] Rotated Bit Packed Glyph Array (Discussion #50) 


@mqshaw<https://github.com/mqshaw> 
How're you and the family been keeping? All well I pray. 

Been working on a STAC variant. The ATOM based device is a bit small to be seen from the stage to the back wall so I'm building a version with a larger display & an external WiFi antenna; code name STAC-LR. Should also be able to incorporate OTA updating and mDNS with this (which I'll back port to the STAC). 

Anyway, larger COTS displays use an 8 x 8 RGB LED matrix instead of the ATOM 5 x 5. I've redone the display routines to accommodate, but hit a bit if a wall regarding the glyph matrix. A bit more detail in the attached sketch, but basically the 8 x 8 glyphs are in bit-packed format. The bitmap for each row is packed into a single byte (uint8_t), for a total of 8 bytes to represent the glyph. 

Wondering if you would have a routine or be able to craft one that would produce the run-time glyph array rotated according to the STAC orientation? 

Merry Christmas! 
Display8.zip<https://github.com/Xylopyrographer/STAC/files/10232630/Display8.zip>

**Comment by mqshaw** [comment]
Hey Rob,

I think that this should give your what you want.  Please not it is 100% untested code 😊


void rotate8x8Glyphs( ORIENTATION stacOrientation ) {
    /*
      Rotates the entire glyph matrix in memory depending upon the vertical orientation of the STAC
    */

    int rotate_0[64] = {0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63} ;
    int rotate_90[64] = {56,48,40,32,24,16,8,0,57,49,41,33,25,17,9,1,58,50,42,34,26,18,10,2,59,51,43,35,27,19,11,3,60,52,44,36,28,20,12,4,61,53,45,37,29,21,13,5,62,54,46,38,30,22,14,6,63,55,47,39,31,23,15,7} ;
    int rotate_180[64] = {63,62,61,60,59,58,57,56,55,54,53,52,51,50,49,48,47,46,45,44,43,42,41,40,39,38,37,36,35,34,33,32,31,30,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1,0} ;
    int rotate_270[64] = {7,15,23,31,39,47,55,63,6,14,22,30,38,46,54,62,5,13,21,29,37,45,53,61,4,12,20,28,36,44,52,60,3,11,19,27,35,43,51,59,2,10,18,26,34,42,50,58,1,9,17,25,33,41,49,57,0,8,16,24,32,40,48,56};

    // Initalize the rotation LUT
    int * rotation_LUT = NULL;
    int tmp_array_in[64] = {0} ;
    int tmp_array_out[64] = {0} ;

    // Determine which rotation to use
    if ( stacOrientation == ORIENTATION::DOWN )                 // Upside Down
      rotation_LUT = rotate_180;
    else if ( stacOrientation == ORIENTATION::LEFT )            // Rotated to left
      rotation_LUT = rotate_90;
    else if ( stacOrientation == ORIENTATION::RIGHT )           // Rotated to Right
      rotation_LUT = rotate_270;
    else
      rotation_LUT = rotate_0;

    // Perform the rotation
    for ( int glyphID = 0 ; glyphID < TOTAL_GLYPHS; glyphID++ )
    {
        int glyph_bitCount = 0 ;

        // Start by unpacking the current Glyph
        for ( int loop_pix = 0 ; loop_pix<8 ; loop_pix++ )
        {
            tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 1 ;
            tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 2 ;
            tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 4 ;
            tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 8 ;
            tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 16 ;
            tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 32 ;
            tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 64 ;
            tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 128 ;
        }

        // Now rotate the Glyph
        for ( int loop_pix = 0; loop_pix < 64; loop_pix++ )
        {
          tmp_array_out[ loop_pix ] = tmp_array_in[ rotation_LUT[ loop_pix ] ] ;
        }

        // Reset the Bit Counter
        glyph_bitCount = 0 ;

        // Finish by repacking the current Glyph
        for ( int loop_pix = 0 ; loop_pix<8 ; loop_pix++ )
        {
            glyphMap[glyphID][ loop_pix ] = tmp_array_out[ glyph_bitCount++ ] >> 0 +
                                            tmp_array_out[ glyph_bitCount++ ] >> 1 +
                                            tmp_array_out[ glyph_bitCount++ ] >> 2 +
                                            tmp_array_out[ glyph_bitCount++ ] >> 3 +
                                            tmp_array_out[ glyph_bitCount++ ] >> 4 +
                                            tmp_array_out[ glyph_bitCount++ ] >> 5 +
                                            tmp_array_out[ glyph_bitCount++ ] >> 6 +
                                            tmp_array_out[ glyph_bitCount++ ] >> 7 ;
        }

    }

}

**Comment by beamholder** [comment]
Hiya Mark

Thanks so much. I’ll give it a go this weekend and let you know 👍. 

Cheers,
    Rob
========================================


On Dec 16, 2022, at 9:09 PM, Mark Shaw ***@***.***> wrote:

﻿
Hey Rob, 

I think that this should give your what you want. Please not it is 100% untested code 😊 


void rotate8x8Glyphs( ORIENTATION stacOrientation ) { 
/* 
Rotates the entire glyph matrix in memory depending upon the vertical orientation of the STAC 
*/ 

int rotate_0[64] = {0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63} ; 
int rotate_90[64] = {56,48,40,32,24,16,8,0,57,49,41,33,25,17,9,1,58,50,42,34,26,18,10,2,59,51,43,35,27,19,11,3,60,52,44,36,28,20,12,4,61,53,45,37,29,21,13,5,62,54,46,38,30,22,14,6,63,55,47,39,31,23,15,7} ; 
int rotate_180[64] = {63,62,61,60,59,58,57,56,55,54,53,52,51,50,49,48,47,46,45,44,43,42,41,40,39,38,37,36,35,34,33,32,31,30,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1,0} ; 
int rotate_270[64] = {7,15,23,31,39,47,55,63,6,14,22,30,38,46,54,62,5,13,21,29,37,45,53,61,4,12,20,28,36,44,52,60,3,11,19,27,35,43,51,59,2,10,18,26,34,42,50,58,1,9,17,25,33,41,49,57,0,8,16,24,32,40,48,56}; 

// Initalize the rotation LUT 
int * rotation_LUT = NULL; 
int tmp_array_in[64] = {0} ; 
int tmp_array_out[64] = {0} ; 

// Determine which rotation to use 
if ( stacOrientation == ORIENTATION::DOWN ) // Upside Down 
rotation_LUT = rotate_180; 
else if ( stacOrientation == ORIENTATION::LEFT ) // Rotated to left 
rotation_LUT = rotate_90; 
else if ( stacOrientation == ORIENTATION::RIGHT ) // Rotated to Right 
rotation_LUT = rotate_270; 
else 
rotation_LUT = rotate_0; 

// Perform the rotation 
for ( int glyphID = 0 ; glyphID < TOTAL_GLYPHS; glyphID++ ) 
{ 
int glyph_bitCount = 0 ; 

// Start by unpacking the current Glyph 
for ( int loop_pix = 0 ; loop_pix<8 ; loop_pix++ ) 
{ 
tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 1 ; 
tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 2 ; 
tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 4 ; 
tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 8 ; 
tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 16 ; 
tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 32 ; 
tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 64 ; 
tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] && 128 ; 
} 

// Now rotate the Glyph 
for ( int loop_pix = 0; loop_pix < 64; loop_pix++ ) 
{ 
tmp_array_out[ loop_pix ] = tmp_array_in[ rotation_LUT[ loop_pix ] ] ; 
} 

// Reset the Bit Counter 
glyph_bitCount = 0 ; 

// Finish by repacking the current Glyph 
for ( int loop_pix = 0 ; loop_pix<8 ; loop_pix++ ) 
{ 
glyphMap[glyphID][ loop_pix ] = tmp_array_out[ glyph_bitCount++ ] >> 0 + 
tmp_array_out[ glyph_bitCount++ ] >> 1 + 
tmp_array_out[ glyph_bitCount++ ] >> 2 + 
tmp_array_out[ glyph_bitCount++ ] >> 3 + 
tmp_array_out[ glyph_bitCount++ ] >> 4 + 
tmp_array_out[ glyph_bitCount++ ] >> 5 + 
tmp_array_out[ glyph_bitCount++ ] >> 6 + 
tmp_array_out[ glyph_bitCount++ ] >> 7 ; 
} 

} 

}

**Comment by mqshaw** [comment]
Great,

If you send me the branch name that you are working on I can check it out and take a look.  I know of one mistake ( I forgot the return )

Mark

**Comment by mqshaw** [comment]
Hi Rob,

A couple more mistakes …

I used the && instead of the & on the following lines ( & is bitwise and and && is logical and ).

      // Start by unpacking the current Glyph
      for ( int loop_pix = 0 ; loop_pix<8 ; loop_pix++ )
      {
          tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] & 1 ;
          tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] & 2 ;
          tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] & 4 ;
          tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] & 8 ;
          tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] & 16 ;
          tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] & 32 ;
          tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] & 64 ;
          tmp_array_in[ glyph_bitCount++ ] = glyphMap[glyphID][ loop_pix ] & 128 ;
      }

And I think that my shift operators are going in the wrong direction  ☹

      // Finish by repacking the current Glyph
      for ( int loop_pix = 0 ; loop_pix<8 ; loop_pix++ )
      {
          glyphMap[glyphID][ loop_pix ] = tmp_array_out[ glyph_bitCount++ ] +
          tmp_array_out[ glyph_bitCount++ ] << 1 +
          tmp_array_out[ glyph_bitCount++ ] << 2 +
          tmp_array_out[ glyph_bitCount++ ] << 3 +
          tmp_array_out[ glyph_bitCount++ ] << 4 +
          tmp_array_out[ glyph_bitCount++ ] << 5 +
          tmp_array_out[ glyph_bitCount++ ] << 6 +
          tmp_array_out[ glyph_bitCount++ ] << 7 ;
      }

Mark

**Reply by Xylopyrographer** [reply]
@mqshaw 
Howdy Mark:

Managed to get it all working 😄!

Took a quite a bit of mental fog-clearing to get the bit packing code figured. Actually resorted to paper and pencil. I blame it on the sudden cold spell.

Did this as a "quick" test routine (ha ha) so haven't rolled it into the base code yet. Attached is what finally came out as working.

All the debug code is still in there so you don't need an actual LED panel connected. The output also dumps to the serial monitor. Target board is a bog standard ESP32 Dev kit unit. Probably wouldn't work on an ATOM Matrix. The "LR" uses a different IMU along with the display differences. But then again the display pin is different and the IMU isn't used so... 

Enjoy looking at your code. I always  find a lesson or two. Or three. Very much appreciated. 🙏🏻

[Display8.zip](https://github.com/Xylopyrographer/STAC/files/10254462/Display8.zip)

**Comment by mqshaw** [comment, marked as answer]
That’s great, I am very glad that you got it working.  I don’t have one of the big boards so it’s always an engineered guess without being able to step through the code in the debugger.

When I look at this section, I realize that I must have got my MSB/LSB order mixed up!  Darn endianness!

        // Finish by repacking the current glyph & saving it into the glyphMap[]
        for ( uint8_t row = 0; row < 8; row++ ) {
            glyphMap[ glyphID ][ row ] =    tmp_array_out[ ( row * 8 ) + 0 ] << 7 |
                                            tmp_array_out[ ( row * 8 ) + 1 ] << 6 |
                                            tmp_array_out[ ( row * 8 ) + 2 ] << 5 |
                                            tmp_array_out[ ( row * 8 ) + 3 ] << 4 |
                                            tmp_array_out[ ( row * 8 ) + 4 ] << 3 |
                                            tmp_array_out[ ( row * 8 ) + 5 ] << 2 |
                                            tmp_array_out[ ( row * 8 ) + 6 ] << 1 |
                                            tmp_array_out[ ( row * 8 ) + 7 ] << 0;
        }

BTW -  We have been running the latest V2 code for about 3 months now and it has been working pretty solidly, so I think that we have reached a good compromise with the tracking code that we added.

Merry Christmas, hope you and your family have a great one.

Best
Mark

**Reply by Xylopyrographer** [reply]
@mqshaw 
Great to hear about STAC v2. The error checking part of the code is much easier to tweak now should we need to.

All the very best to you and the family this Christmas.
